syntax = "proto3";

message InputSeatBlockSetup {
    repeated CustomAxisSetup CustomAxes = 1;
    bool LookAroundDisabled = 2;
}

message SliderBlockSetup {
    float sliderValue = 1;
    float sliderValue2 = 2;
    float sliderValue3 = 3;
    float sliderValue4 = 4;
    reserved 5;  // TODO: ToneBlockSetup handling
}

message ToneBlockSetup {
    int32 sampleIndex = 1;
}

message VisitedStateMap {
    repeated bool states = 1;
    SimpleBitArray bitMaskedStates = 2;
}

message Vector3 {
    float x = 1;
    float y = 2;
    float z = 3;
}

message Quaternion {
    float x = 1;
    float y = 2;
    float z = 3;
    float w = 4;
}

message Bounds {
    Vector3 center = 1;
    Vector3 size = 2;
}

message Rect {
    float x = 1;
    float y = 2;
    float width = 3;
    float height = 4;
}

message StructureGraphSaveDataProto {
    repeated StructureNodeSaveData StructureNodes = 1;
    repeated int32 Edges = 2;
    Bounds Bounds = 3;
    float LowestPoint = 4;
    Vector3 MidPointOffsetFromWeldGroupSpace = 5;
    string oldGameVersion = 6;
    int32 SaveDataVersion = 7;
    GameVersion GameVersion = 8;
    reserved 9;
    WeldingDataProto WeldingData = 10;
}

message BlockBehaviourSetup {
    repeated SocketLinkHandleDescription LinkHandlePositiveList = 1;
    repeated SocketLinkHandleDescription LinkHandleNegativeList = 2;
    reserved 3;  // TODO: InputSeatBlockSetup handling
    SerializableColor primaryColor = 4;
    SerializableColor secondaryColor = 5;
    reserved 6;  // TODO: SliderBlockSetup handling
    float DelayTime = 7;
    float ActiveTime = 8;
    bool TogglePositive = 9;
    bool ToggleNegative = 10;
    float InActiveTime = 11;
    KeyCode NegativeKeyOld = 12;
    KeyCode PositiveKeyOld = 13;
    EAction NegativeAction = 14;
    EAction PositiveAction = 15;
    bool CustomToggle = 16;
    BlockSkinType SkinType = 17;
}

message GhostData {
    repeated SerializableVector3 posPerTimeUnit = 1;
    repeated SerializableVector3 rotPerTimeUnit = 2;
    repeated float timeStampPerTimeUnit = 3;
    reserved 4;  // TODO: array StructureSaveData
    float timeScale = 5;
    float raceTime = 6;
    int32 lastIndex = 7;
}

message WeldingDataProto {
    repeated KeyValuePairIntListInt m_weldgroupsDict = 1;
    repeated int32 m_weldEdgesList = 2;
    int32 WeldSettingsHash = 3;
}

message ListInt {
    repeated int32 value = 1;
}

message TrailmakersPlayerSaveData {
    BlockUnlockingSystemSaveData m_blockUnlockingData = 1;
    GoldPlayerSystemSaveData m_goldData = 2;
    BuildPowerSystemSaveData m_buildPowerData = 3;
    CheckpointSystemSaveData m_checkpointData = 4;
    reserved 5;
    ShroudSystemSaveData m_shroudData = 6;
    UnlockingPlayerSystemSaveData m_unlockingData = 7;
    SecretSystemSaveData m_secretData = 8;
    reserved 10;  // TODO: array m_latestStructureData
    StatisticsPlayerSystemSaveData m_statisticsData = 11;
    TreasureHuntSystemSaveData m_treasureHuntData = 12;
    NuggetHuntSystemSaveData m_nuggetHuntData = 13;
    SalvagePlayerSystemSaveData m_salvageData = 14;
    reserved 15;  // TODO: array latestStructureIdentifierBytes
}

message StructureSaveIdentifierProto {
    EIdentifierType IdentifierType = 1;
    string Identifier = 2;
}

message StructureMetaDataProto {
    string Name = 1;
    string Description = 2;
    repeated string Tags = 3;
    string CreatorName = 4;
    string CreatorID = 5;
    string Platform = 6;
}

message TransformationLinksProto {
    repeated StructureSaveIdentifierProto Links = 1;
}

message Guid {
    uint64 low = 1;
    uint64 high = 2;
}

message StructureNodeSaveData {
    ABuildGridPosition SocketPosition = 1;
    ABuildGridRotation SocketRotation = 2;
    int32 SocketIndex = 3;
    int32 parentSocket = 4;
    int32 StructureIndex = 5;
    BlockBehaviourSetup BlockSetup = 6;
    Guid BlockMetadataGuid = 7;
    reserved 8;
    int32 CloneNodeID = 9;
}

message GameVersion {
    int32 GameVersion_Major = 1;
    int32 GameVersion_Mid = 2;
    int32 GameVersion_Minor = 3;
    int32 GameVersion_ChangeSet = 4;
}

message SocketLinkHandleDescription {
    int32 m_structureOffset = 1;
    int32 m_guidString = 2;
    int32 m_socketGUID = 3;  // TODO: `int?`
    int32 LinkHandleIndex = 4;
    KeyCode OldNegativeKey = 5;
    KeyCode OldPositiveKey = 6;
    int32 m_socketOffset = 7;
}

message SerializableColor {
    float r = 1;
    float g = 2;
    float b = 3;
    float a = 4;
    int32 m_commonColorIndexOffByOne = 5;
}

enum KeyCode {
    // For Command/Apple keys
    option allow_alias = true;

    None = 0;
    Backspace = 8;
    Tab = 9;
    Clear = 12;
    Return = 13;
    Pause = 19;
    Escape = 27;
    Space = 32;
    Exclaim = 33;
    DoubleQuote = 34;
    Hash = 35;
    Dollar = 36;
    Percent = 37;
    Ampersand = 38;
    Quote = 39;
    LeftParen = 40;
    RightParen = 41;
    Asterisk = 42;
    Plus = 43;
    Comma = 44;
    Minus = 45;
    Period = 46;
    Slash = 47;
    Alpha0 = 48;
    Alpha1 = 49;
    Alpha2 = 50;
    Alpha3 = 51;
    Alpha4 = 52;
    Alpha5 = 53;
    Alpha6 = 54;
    Alpha7 = 55;
    Alpha8 = 56;
    Alpha9 = 57;
    Colon = 58;
    Semicolon = 59;
    Less = 60;
    Equals = 61;
    Greater = 62;
    Question = 63;
    At = 64;
    LeftBracket = 91;
    Backslash = 92;
    RightBracket = 93;
    Caret = 94;
    Underscore = 95;
    BackQuote = 96;
    A = 97;
    B = 98;
    C = 99;
    D = 100;
    E = 101;
    F = 102;
    G = 103;
    H = 104;
    I = 105;
    J = 106;
    K = 107;
    L = 108;
    M = 109;
    N = 110;
    O = 111;
    P = 112;
    Q = 113;
    R = 114;
    S = 115;
    T = 116;
    U = 117;
    V = 118;
    W = 119;
    X = 120;
    Y = 121;
    Z = 122;
    LeftCurlyBracket = 123;
    Pipe = 124;
    PightCurlyBracket = 125;
    Tilde = 126;
    Delete = 127;
    Keypad0 = 256;
    Keypad1 = 257;
    Keypad2 = 258;
    Keypad3 = 259;
    Keypad4 = 260;
    Keypad5 = 261;
    Keypad6 = 262;
    Keypad7 = 263;
    Keypad8 = 264;
    Keypad9 = 265;
    KeypadPeriod = 266;
    KeypadDivide = 267;
    KeypadMultiply = 268;
    KeypadMinus = 269;
    KeypadPlus = 270;
    KeypadEnter = 271;
    KeypadEquals = 272;
    UpArrow = 273;
    DownArrow = 274;
    RightArrow = 275;
    LeftArrow = 276;
    Insert = 277;
    Home = 278;
    End = 279;
    PageUp = 280;
    PageDown = 281;
    F1 = 282;
    F2 = 283;
    F3 = 284;
    F4 = 285;
    F5 = 286;
    F6 = 287;
    F7 = 288;
    F8 = 289;
    F9 = 290;
    F10 = 291;
    F11 = 292;
    F12 = 293;
    F13 = 294;
    F14 = 295;
    F15 = 296;
    Numlock = 300;
    CapsLock = 301;
    ScrollLock = 302;
    RightShift = 303;
    LeftShift = 304;
    RightControl = 305;
    LeftControl = 306;
    RightAlt = 307;
    LeftAlt = 308;
    RightCommand = 309;
    RightApple = 309;
    LeftCommand = 310;
    LeftApple = 310;
    LeftWindows = 311;
    RightWindows = 312;
    AltGr = 313;
    Help = 315;
    Print = 316;
    SysReq = 317;
    Break = 318;
    Menu = 319;
    Mouse0 = 323;
    Mouse1 = 324;
    Mouse2 = 325;
    Mouse3 = 326;
    Mouse4 = 327;
    Mouse5 = 328;
    Mouse6 = 329;
    JoystickButton0 = 330;
    JoystickButton1 = 331;
    JoystickButton2 = 332;
    JoystickButton3 = 333;
    JoystickButton4 = 334;
    JoystickButton5 = 335;
    JoystickButton6 = 336;
    JoystickButton7 = 337;
    JoystickButton8 = 338;
    JoystickButton9 = 339;
    JoystickButton10 = 340;
    JoystickButton11 = 341;
    JoystickButton12 = 342;
    JoystickButton13 = 343;
    JoystickButton14 = 344;
    JoystickButton15 = 345;
    JoystickButton16 = 346;
    JoystickButton17 = 347;
    JoystickButton18 = 348;
    JoystickButton19 = 349;
    Joystick1Button0 = 350;
    Joystick1Button1 = 351;
    Joystick1Button2 = 352;
    Joystick1Button3 = 353;
    Joystick1Button4 = 354;
    Joystick1Button5 = 355;
    Joystick1Button6 = 356;
    Joystick1Button7 = 357;
    Joystick1Button8 = 358;
    Joystick1Button9 = 359;
    Joystick1Button10 = 360;
    Joystick1Button11 = 361;
    Joystick1Button12 = 362;
    Joystick1Button13 = 363;
    Joystick1Button14 = 364;
    Joystick1Button15 = 365;
    Joystick1Button16 = 366;
    Joystick1Button17 = 367;
    Joystick1Button18 = 368;
    Joystick1Button19 = 369;
    Joystick2Button0 = 370;
    Joystick2Button1 = 371;
    Joystick2Button2 = 372;
    Joystick2Button3 = 373;
    Joystick2Button4 = 374;
    Joystick2Button5 = 375;
    Joystick2Button6 = 376;
    Joystick2Button7 = 377;
    Joystick2Button8 = 378;
    Joystick2Button9 = 379;
    Joystick2Button10 = 380;
    Joystick2Button11 = 381;
    Joystick2Button12 = 382;
    Joystick2Button13 = 383;
    Joystick2Button14 = 384;
    Joystick2Button15 = 385;
    Joystick2Button16 = 386;
    Joystick2Button17 = 387;
    Joystick2Button18 = 388;
    Joystick2Button19 = 389;
    Joystick3Button0 = 390;
    Joystick3Button1 = 391;
    Joystick3Button2 = 392;
    Joystick3Button3 = 393;
    Joystick3Button4 = 394;
    Joystick3Button5 = 395;
    Joystick3Button6 = 396;
    Joystick3Button7 = 397;
    Joystick3Button8 = 398;
    Joystick3Button9 = 399;
    Joystick3Button10 = 400;
    Joystick3Button11 = 401;
    Joystick3Button12 = 402;
    Joystick3Button13 = 403;
    Joystick3Button14 = 404;
    Joystick3Button15 = 405;
    Joystick3Button16 = 406;
    Joystick3Button17 = 407;
    Joystick3Button18 = 408;
    Joystick3Button19 = 409;
    Joystick4Button0 = 410;
    Joystick4Button1 = 411;
    Joystick4Button2 = 412;
    Joystick4Button3 = 413;
    Joystick4Button4 = 414;
    Joystick4Button5 = 415;
    Joystick4Button6 = 416;
    Joystick4Button7 = 417;
    Joystick4Button8 = 418;
    Joystick4Button9 = 419;
    Joystick4Button10 = 420;
    Joystick4Button11 = 421;
    Joystick4Button12 = 422;
    Joystick4Button13 = 423;
    Joystick4Button14 = 424;
    Joystick4Button15 = 425;
    Joystick4Button16 = 426;
    Joystick4Button17 = 427;
    Joystick4Button18 = 428;
    Joystick4Button19 = 429;
    Joystick5Button0 = 430;
    Joystick5Button1 = 431;
    Joystick5Button2 = 432;
    Joystick5Button3 = 433;
    Joystick5Button4 = 434;
    Joystick5Button5 = 435;
    Joystick5Button6 = 436;
    Joystick5Button7 = 437;
    Joystick5Button8 = 438;
    Joystick5Button9 = 439;
    Joystick5Button10 = 440;
    Joystick5Button11 = 441;
    Joystick5Button12 = 442;
    Joystick5Button13 = 443;
    Joystick5Button14 = 444;
    Joystick5Button15 = 445;
    Joystick5Button16 = 446;
    Joystick5Button17 = 447;
    Joystick5Button18 = 448;
    Joystick5Button19 = 449;
    Joystick6Button0 = 450;
    Joystick6Button1 = 451;
    Joystick6Button2 = 452;
    Joystick6Button3 = 453;
    Joystick6Button4 = 454;
    Joystick6Button5 = 455;
    Joystick6Button6 = 456;
    Joystick6Button7 = 457;
    Joystick6Button8 = 458;
    Joystick6Button9 = 459;
    Joystick6Button10 = 460;
    Joystick6Button11 = 461;
    Joystick6Button12 = 462;
    Joystick6Button13 = 463;
    Joystick6Button14 = 464;
    Joystick6Button15 = 465;
    Joystick6Button16 = 466;
    Joystick6Button17 = 467;
    Joystick6Button18 = 468;
    Joystick6Button19 = 469;
    Joystick7Button0 = 470;
    Joystick7Button1 = 471;
    Joystick7Button2 = 472;
    Joystick7Button3 = 473;
    Joystick7Button4 = 474;
    Joystick7Button5 = 475;
    Joystick7Button6 = 476;
    Joystick7Button7 = 477;
    Joystick7Button8 = 478;
    Joystick7Button9 = 479;
    Joystick7Button10 = 480;
    Joystick7Button11 = 481;
    Joystick7Button12 = 482;
    Joystick7Button13 = 483;
    Joystick7Button14 = 484;
    Joystick7Button15 = 485;
    Joystick7Button16 = 486;
    Joystick7Button17 = 487;
    Joystick7Button18 = 488;
    Joystick7Button19 = 489;
    Joystick8Button0 = 490;
    Joystick8Button1 = 491;
    Joystick8Button2 = 492;
    Joystick8Button3 = 493;
    Joystick8Button4 = 494;
    Joystick8Button5 = 495;
    Joystick8Button6 = 496;
    Joystick8Button7 = 497;
    Joystick8Button8 = 498;
    Joystick8Button9 = 499;
    Joystick8Button10 = 500;
    Joystick8Button11 = 501;
    Joystick8Button12 = 502;
    Joystick8Button13 = 503;
    Joystick8Button14 = 504;
    Joystick8Button15 = 505;
    Joystick8Button16 = 506;
    Joystick8Button17 = 507;
    Joystick8Button18 = 508;
    Joystick8Button19 = 509;
}

enum EAction {
    NoneAction = 0;
    CharacterForward = 1;
    CharacterBack = 2;
    CharacterLeft = 3;
    CharacterRight = 4;
    CharacterJump = 5;
    StructureForward = 6;
    StructureBack = 7;
    StructureLeft = 8;
    StructureRight = 9;
    StructureUp = 10;
    StructureDown = 11;
    Boost = 12;
    AltBoost = 13;
    Shoot = 14;
    AltShoot = 15;
    Extend = 16;
    Compress = 17;
    RotateCW = 18;
    RotateCCW = 19;
    Detch = 20;
    Brake = 21;
    Custom1 = 22;
    Custom2 = 23;
    Custom3 = 24;
    Custom4 = 25;
    Custom5 = 26;
    Custom6 = 27;
    Custom7 = 28;
    Custom8 = 29;
    Custom9 = 30;
    Custom10 = 31;
    Custom11 = 32;
    Custom12 = 33;
    Custom13 = 34;
    Custom14 = 35;
    Custom15 = 36;
    Custom16 = 37;
    Custom17 = 38;
    Custom18 = 39;
    Custom19 = 40;
    Custom20 = 41;
    Custom21 = 42;
    Custom22 = 43;
    Custom23 = 44;
    Custom24 = 45;
    Custom25 = 46;
    Custom26 = 47;
    Custom27 = 48;
    Custom28 = 49;
    Custom29 = 50;
    Custom30 = 51;
    Custom31 = 52;
    Custom32 = 53;
    Custom33 = 54;
    Custom34 = 55;
    Custom35 = 56;
    Custom36 = 57;
    Custom37 = 58;
    Custom38 = 59;
    ClearInput = 60;
    AimLeft = 61;
    AimRight = 62;
    AimUp = 63;
    AimDown = 64;
    MouseLeft = 65;
    MouseRight = 66;
    AirControlForward = 67;
    AirControlBack = 68;
    AirControlLeft = 69;
    AirControlRight = 70;
    Respawn = 71;
    Repair = 72;
    EnterBuild = 73;
    ExitBuild = 74;
    EnterSeat = 75;
    ExitSeat = 76;
    Pickup = 77;
    AcceptOffer = 78;
    Modifier = 79;
    EnterTransform = 80;
    OpenMenu = 81;
    CameraCycle = 82;
    AlternateCamera = 83;
    Chat = 84;
    MapToggle = 85;
    MapExit = 86;
    MapZoomOut = 87;
    MapZoomIn = 88;
    MapTranslateUp = 89;
    MapTranslateDown = 90;
    MapTranslateLeft = 91;
    MapTranslateRight = 92;
    MapCancel = 93;
    Transform0 = 94;
    Transform1 = 95;
    Transform2 = 96;
    Transform3 = 97;
    Transform4 = 98;
    Transform5 = 99;
    Transform6 = 100;
    Transform7 = 101;
    BuilderCursorX_plus = 102;
    BuilderCursorX_minus = 103;
    BuilderCursorY_plus = 104;
    BuilderCursorY_minus = 105;
    BuilderCamRotX_plus = 106;
    BuilderCamRotX_minus = 107;
    BuilderCamRotY_plus = 108;
    BuilderCamRotY_minus = 109;
    BuilderX_plus = 110;
    BuilderX_minus = 111;
    BuilderY_plus = 112;
    BuilderY_minus = 113;
    BuilderActivate = 114;
    BuilderCherryPick = 115;
    BuilderUseDepth = 116;
    BuilderModSelection = 117;
    BuilderCancel = 118;
    BuilderRotateMode = 119;
    BuilderBlueprintContextMenu = 120;
    BuilderClone = 121;
    BuilderDelete = 122;
    BuilderUploadBlueprint = 123;
    BuilderDeleteBlueprint = 124;
    BuilderInstructionToggleButton = 125;
    BuilderTriggerLeft = 126;
    BuilderTriggerRight = 127;
    BuilderShowMetaActionMenu = 128;
    BuilderGotoBlockButtons = 129;
    BuilderUndo = 130;
    BuilderLeaveSubstate = 131;
    BuilderMoveCamera = 132;
    BuilderDragRotateCamera = 133;
    CloseMenu = 134;
    Zoom = 135;
    QuickLoad = 136;
    ToggleHints = 137;
    ExitTransformationState = 138;
    RespawnInstant = 139;
    INVALID = 140;
    BuilderCancelInputAssignment = 141;
    ReleaseCursor = 142;
    BuilderEditMetaData = 143;
    NumActions = 144;
}

enum BlockSkinType {
    NoneSkinType = 0;
    Wood = 1;
    Glow = 2;
    Stone = 3;
    Camouflage = 4;
    Chrome = 5;
    Glass = 6;
    DiagonalA = 7;
    DiagonalB = 8;
    Greeble = 9;
    Scales = 10;
    NonSlip = 11;
    Grate = 12;
    Flesh = 13;
    Length = 14;
}

message CustomAxisSetup {
    KeyCode OldPositiveButton = 1;
    KeyCode OldNegativeButton = 2;
    EAction PositiveButton = 3;
    EAction NegativeButton = 4;
}

message SerializableVector3 {
    float x = 1;
    float y = 2;
    float z = 3;
}

message KeyValuePairIntListInt {
    int32 Key = 1;
    repeated int32 Value = 2;
}

message BlockUnlockingSystemSaveData {
    repeated string UnlockedBlocks = 1;
}

message GoldPlayerSystemSaveData {
    int32 m_value = 1;
    int32 m_max = 2;
    repeated string m_unlockedIds = 3;
}

message BuildPowerSystemSaveData {
    int32 m_power = 1;
    int32 m_maxPower = 2;
    repeated string unlockedScannableIds = 3;
}

message CheckpointSystemSaveData {
    repeated string m_unlockedCheckpointIds = 1;
    string m_currentCheckpointId = 2;
    string m_currentSpawnPointId = 3;
    string m_currentGameplayIslandId = 4;
    string m_highestGamePlayIslandReached = 5;
}

message ShroudSystemSaveData {
    reserved 1;  // TODO: VisitedStateMap (CellMap) VistedMap
}

message UnlockingPlayerSystemSaveData {
    repeated string m_unlocks = 1;
}

message SecretSystemSaveData {
    int32 m_count = 1;
    repeated string m_unlockedIds = 2;
}

message StatisticsPlayerSystemSaveData {
    float m_playtime = 1;
    float m_timeOfDayPercentage = 2;
}

message TreasureHuntSystemSaveData {
    repeated string m_unlockedIds = 1;
}

message NuggetHuntSystemSaveData {
    repeated string m_unlockedIds = 1;
}

message SalvagePlayerSystemSaveData {
    repeated string m_detectedIds = 1;
}

enum EIdentifierType {
    NoIdentifier = 0;
    NotSaved = 1;
    StructureSaveGUID = 2;
    GalleryGUID = 3;
    WorkshopGUID = 4;
    LocalFilePathHACK_StructureSave = 5;
    LocalFilePathHACK_AutoSave = 6;
    LocalFilePathHACK_LegacyTransformationSlot = 7;
    PremadeVehicleHACK = 8;
}

message ABuildGridPosition {
    // All of those should really be short/int16...
    int32 X = 1;
    int32 Y = 2;
    int32 Z = 3;
}

message ABuildGridRotation {
    // All of those should really be short/int16...
    int32 EulerX = 1;
    int32 EulerY = 2;
    int32 EulerZ = 3;
}

message CellMap {
    int32 width = 1;
    int32 height = 2;
    Rect bounds = 3;
    reserved 4;  // TODO: VisitedStateMap handling
}

message SimpleBitArray {
    repeated int32 m_array = 1;
    int32 m_Length = 2;
}
